shoop/core/cache.py

```diff
@@ -0,0 +1,161 @@
+# -*- coding: utf-8 -*-
+"""
+Utilities for caching, with versioned caching and automatic timeout determination.
+
+Versioning works by way of "namespaces". Namespaces are the first colon-separated part
+of cache keys.
+
+For instance, the cache keys `price:10`, `price:20`, and `price` all belong to the
+`price` namespace and can be invalidated with one `bump_version("price")` call.
+
+The versions themselves are stored within the cache, within the `_version` namespace.
+(As an implementation detail, this allows one to invalidate _all_ versioned keys by
+bumping the version of `_version`. Very meta!)
+"""
+import time
+from django.conf import settings
+from django.utils.encoding import force_text
+from django.core.cache import caches
+from django.core.signals import request_finished
+from threading import local
+
+DEFAULT_CACHE_DURATIONS = {}
+_versions = local()
+
+
+def _clear_versions_for_request(**kwargs):
+    _versions.__dict__.clear()
+
+
+request_finished.connect(_clear_versions_for_request, dispatch_uid="shoop.core.cache._clear_versions_for_request")
+
+
+def _get_cache_key_namespace(cache_key):
+    """
+    Split the given cache key by the first colon to get a namespace.
+
+    :param cache_key: Cache key string
+    :type cache_key: str
+    :return: Cache namespace string
+    :rtype: str
+    """
+    return force_text(cache_key).split(":")[0]
```

Comment:
I'd rather `assert isinstance(cache_key, six.text_type)` than force_text it to avoid unneeded UTF-8/unicode conversions. API caller should know that the API expects text rather than binary and use it with str (or on py2: unicode) rather than bytes.

And, `split(":", 1)[0]`, since we don't need to split more than once.